### ISSUE TITLE
ENV variables not loaded with config

### ISSUE THREAD

**Issue by deepakmahakale** [rubyconfig/config#321]
I followed [this](https://github.com/rubyconfig/config#fine-tuning) but I am unable to access the environment variable with `Settings`

config.rb
```ruby
Config.setup do |config|
  config.use_env = true
  config.env_prefix = 'SETTINGS'
  config.env_separator = '__'
  config.env_converter = :downcase
  config.env_parse_values = true
end
```

```console
SETTINGS__SECTION__SERVER_SIZE=1 SETTINGS__SECTION__SERVER=google.com SETTINGS__SECTION__SSL_ENABLED=false rails c
```

```ruby
Settings.section.server_size # => nil
Settings.section.server      # => nil
Settings.section.ssl_enabled # => nil

# They are accessible this way
ENV['SETTINGS__SECTION__SERVER_SIZE'] #=> "1"
ENV['SETTINGS__SECTION__SERVER']      #=> "google.com"
ENV['SETTINGS__SECTION__SSL_ENABLED'] #=> "false"
```

  * ruby 3.0.4p208 (2022-04-12 revision 3fa771dded) [x86_64-darwin20]
  * rails (6.1.6)
  * config (4.0.0)

**Comment by harrisjb** [comment]
Similar issue here:
- config 4.0.0
- ruby 2.5.8
- rails 5.1.7


Contents of `config/initializers/config.rb`
```ruby
# set up config to look for ENV[key] instead of Settings.key
Config.setup do |config|
  config.use_env = true
end    
```

rails console does not load into ENV namespace:
```ruby
$ rails c
Loading development environment (Rails 5.1.7)
[1] pry(main)> ENV['database_host']
=> nil
[2] pry(main)> Settings.database_host
=> "localhost"
```

**Comment by harrisjb** [comment]
After review I believe this is expected behavior as the gem is just trying to read from `ENV` namespace if things are loaded into `ENV` by Heroku or AWS or some other means. It doesn't intend to write environment variables into the `ENV` namespace itself, it's just reading from it.

**Comment by pkuczynski** [comment]
Yes @harrisjb we do not populate things back into ENV, we only read from it. So your reported issue is not valid.

**Comment by pkuczynski** [comment]
However what @deepakmahakale posted should work. @cjlarose any idea why it does not?

**Comment by jfadgen** [comment]
I am having the same issue as @deepakmahakale , in a new Rails app:

* config 4.0.0
* rails 7.0.3
* ruby 2.7.4

```
Config.setup do |config|
  config.use_env = true
  config.env_prefix = 'CONFIG'
  config.env_separator = '__'
  config.env_converter = :downcase
  config.env_parse_values = true
end
```

.env
```
CONFIG__COGNITO__USER_POOL_ID="test_pool_id"
CONFIG__COGNITO__CLIENT_ID="test_client_id"
CONFIG__COGNITO__CLIENT_SECRET_KEY="test_secret_key"
CONFIG__ANYTHING_TEST="anything at all"
```


```
irb(main):001:0> ENV["CONFIG__COGNITO__USER_POOL_ID"]
=> "test_pool_id"
irb(main):002:0> ENV["CONFIG__COGNITO__CLIENT_ID"]
=> "test_client_id"
irb(main):003:0> ENV["CONFIG__COGNITO__CLIENT_SECRET_KEY"]
=> "test_secret_key"
irb(main):004:0> ApplicationConfig.cognito
=> nil
irb(main):005:0> ENV["CONFIG__ANYTHING_TEST"]
=> "anything at all"
irb(main):006:0> ApplicationConfig.anything_test
=> nil
```

Happy to try anything if you have any recommendations, thank you.

**Comment by nick-reshetnikov** [comment]
@jfadgen, @deepakmahakale, I ran into a similar issue in a recent rails project and figured out that the problem is that ENV variables haven't been loaded yet when the config initializer runs if you're using `.env` files.

specifically, if you're loading from a `.env` file, you will likely need load the .env files before the initializers run.

according to https://github.com/bkeepers/dotenv:
> dotenv is initialized in your Rails app during the before_configuration callback, which is fired when the Application constant is defined in config/application.rb with class Application < Rails::Application. If you need it to be initialized sooner, you can manually call Dotenv::Railtie.load.
```
# config/application.rb
Bundler.require(*Rails.groups)

# Load dotenv only in development or test environment
if ['development', 'test'].include? ENV['RAILS_ENV']
  Dotenv::Railtie.load
end
```

note that if you're using something like heroku, where the env vars are set on the servers, they will likely be available for the initializers.  so you may only need to do this for your dev and test environments.

@pkuczynski I think you guys should think about adding support for .env files - if they're detected, your gem should automatically try to load the contents of `.env` before `config/initializers/config.rb` runs

**Comment by pkuczynski** [comment]
The whole concept of this gem is to store config in the per env configuration files. Using .env and gems like dotenv is a different approach for doing the same thing. They kind of contradictory to each other. So I am not sure if we should support this. @cjlarose what do you think?

**Comment by nick-reshetnikov** [comment]
yeah, I see what you're saying, but the issue is that a lot of teams use `.env` files for dev/test environments, but manage environment variables directly on staging/prod servers (e.g. thru Heroku config vars).  pretty much every project I've worked on has used this hybrid approach.  I honestly haven't seen `.env.production` really used, certainly not with things like heroku.

so, you're kind of leaving the door open to a lot of confusion if you don't support env vars being preloaded thru something like dotenv, which is a very popular tool.

for me, I wanted to use this gem primarily to store certain configs in a yml file format, which allows for nesting keys and other advantages over long, messy .env files.  however, I also wanted to be able to override default values in that yml file thru env vars (i.e. on heroku).  I think that's a legitimate use case.


but in any case, you guys should at least add a disclaimer about this scenario in the README, since I'm sure others have ran into this issue

**Comment by jfadgen** [comment]
Worth noting that the issue went away for me after changing my Gemfile to bundle 
```gem "dotenv-rails"``` 
instead of 
```gem "dotenv"```.

**Comment by nick-reshetnikov** [comment]
@jfadgen I in fact saw this issue while using `"dotenv-rails"`

**Comment by nick-reshetnikov** [comment]
actually, I just discovered that you can require `dotenv/rails-now` (https://github.com/bkeepers/dotenv/blob/master/lib/dotenv/rails-now.rb) in your gemfile to solve the issue:
```
gem 'dotenv-rails', require: 'dotenv/rails-now'
```
this will ensure that dotenv files are loaded before initializers and solve this issue.  I think modifying the gemfile is preferable to messing with `application.rb` in the manner I showed above.

**Comment by kmordan24** [comment]
For anyone else coming into this issue and seeing @nick-reshetnikov 's solution, you'll get a warning w/ latest version that `dotenv/rails-now` is deprecated and instead to use `dotenv/load`.

`gem "dotenv-rails", require: "dotenv/load"`
